src/sentry/scripts/digests/digests.lua

```diff
@@ -0,0 +1,318 @@
+        -- should always exist, and we established that above) but not
+        -- using that here doesn't break anything and allows use to use
+        -- older Redis versions.
+        redis.call('ZADD', configuration:get_schedule_waiting_key(), update, timeline_id)
+    end
+
+    return false
+end
+
+local function truncate_timeline(configuration, timeline_id, capacity)
+    return zset_trim(
+        configuration:get_timeline_key(timeline_id),
+        capacity,
+        function (record_id)
+            redis.call('DEL', configuration:get_timeline_record_key(timeline_id, record_id))
+        end
+    )
+end
+
+local function truncate_digest(configuration, timeline_id, capacity)
+    return zset_trim(
+        configuration:get_timeline_digest_key(timeline_id),
+        capacity,
+        function (record_id)
+            redis.call('DEL', configuration:get_timeline_record_key(timeline_id, record_id))
+        end
+    )
+end
+
+local function add_record_to_timeline(configuration, timeline_id, record_id, value, timestamp, delay_increment, delay_maximum, timeline_capacity, truncation_chance)
+    redis.call('SETEX', configuration:get_timeline_record_key(timeline_id, record_id), configuration.ttl, value)
+    redis.call('ZADD', configuration:get_timeline_key(timeline_id), timestamp, record_id)
+    redis.call('EXPIRE', configuration:get_timeline_key(timeline_id), configuration.ttl)
+
+    local ready = add_timeline_to_schedule(configuration, timeline_id, timestamp, delay_increment, delay_maximum)
+
+    -- TODO: `tonumber` should happen upstream
+    local timeline_capacity = tonumber(timeline_capacity)
+    if timeline_capacity > 0 and math.random() < tonumber(truncation_chance) then
+        -- TODO: `tonumber` should happen upstream
+        truncate_timeline(configuration, timeline_id, timeline_capacity)
+    end
+
+    return ready
+end
+
+local function digest_timeline(configuration, timeline_id)
+    -- Check to ensure that the timeline is in the correct state.
+    if redis.call('ZSCORE', configuration:get_schedule_ready_key(), timeline_id) == false then
+        error('err(invalid_state): timeline is not in the ready state, cannot be digested')
+    end
+
+    local digest_key = configuration:get_timeline_digest_key(timeline_id)
+    local timeline_key = configuration:get_timeline_key(timeline_id)
+    if redis.call('EXISTS', digest_key) == 1 then
+        -- If the digest set already exists (possibly because we already tried
+        -- to send it and failed for some reason), merge any new data into it.
+        -- TODO: It might make sense to trim here to avoid returning capacity *
+        -- 2 if timeline was full when it was previously digested.
+        redis.call('ZUNIONSTORE', digest_key, 2, timeline_key, digest_key, 'AGGREGATE', 'MAX')
+        redis.call('DEL', timeline_key)
+        redis.call('EXPIRE', digest_key, configuration.ttl)
+    elseif redis.call('EXISTS', timeline_key) == 1 then
```

Comment:
We didn't make that assumption — `ZUNIONSTORE` and `DEL` will safely handle missing source keys, so if those don't exist those two commands will still work OK.

This chunk here (starting at L191) is just a way to move the timeline contents to the digest key and make sure that we don't accidentally clobber any data at the digest key in the process. This bit starting at L201 is just a quicker way to do that if the digest key doesn't already exist, since `RENAME` is `O(1)` (but will implicitly delete any data at the destination key if it exists!) and `ZUNIONSTORE` is `O(N)`(-ish.)

I guess it would actually be more efficient to restructure it so that:

- If the digest set exists and the timeline set exists, merge the timeline set into the digest set and delete the timeline set.
- If the digest set exists and the timeline set doesn't exist, do nothing. (This is implicitly what is happening in the first block right now since `ZUNIONSTORE` and `DEL` have no effect with the missing key.)
- If the digest set doesn't exist and the timeline set exists, move the timeline set to the digest set key.
- If both sets don't exist, do nothing.

I guess in pseudo-code, that'd look like:

```lua
if the timeline key exists then
  if the digest key exists then
    -- Merge timeline set into digest set (overwriting the digest set in the process.)
    -- Delete timeline set.
  else
    -- Move timeline set to digest set key.
  end
  -- Set TTL on the digest set.
end
```